### PR TITLE
Fixed issue #18071: Importing Question through Remote Control may fai…

### PR DESCRIPTION
**Pull request by Shnoulle**
Fixed issue #18071: Importing Question through Remote Control may fail if the imported question already exists
Dev: check title unicity before import
Dev: set autorename to tru if new title is set
Dev: review permission (sureycontent/import or update)
Fixed issue #16907: Importing Question through Remote Control doesn't set new title for all languages
Fixed issue #18070: Importing Question through Remote Control doesn't set text for all languages
Dev: most biggest issue : title broken too :)
Dev: set for all language and return error if not validate

### CODE REVIEW

**Comment by Shnoulle** [comment]
I create the 5.X fix after , but need different fixes

**Review by gabrieljenik — APPROVED** [review]
Code structure looks OK. Havne;t tested

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 1]
```diff
@@ -1393,104 +1393,118 @@ public function delete_question($sSessionKey, $iQuestionID)
...
+        $count = 0;
+        foreach ($oQuestions as $oQuestion) {
+            $count++;
+            if (!empty($sNewQuestionTitle)) {
+                $oQuestion->setAttribute('title', $sNewQuestionTitle);
+            }
+            if (!empty($sNewqQuestion)) {
+                $oQuestion->setAttribute('question', $sNewqQuestion);
+            }
+            if (!empty($sNewQuestionHelp)) {
```
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/helpers/remotecontrol/remotecontrol_handle.php:1488]: Yes, same fix for questintext than help

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 2]
```diff
@@ -1393,104 +1393,118 @@ public function delete_question($sSessionKey, $iQuestionID)
...
+            $sMandatory = 'N';
+        }
+        $errors = [];
+        $count = 0;
+        foreach ($oQuestions as $oQuestion) {
+            $count++;
+            if (!empty($sNewQuestionTitle)) {
+                $oQuestion->setAttribute('title', $sNewQuestionTitle);
+            }
+            if (!empty($sNewqQuestion)) {
```
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/helpers/remotecontrol/remotecontrol_handle.php:1485]: Yes : 18070 is in the comment.

File: application/helpers/remotecontrol/remotecontrol_handle.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/helpers/remotecontrol/remotecontrol_handle.php:1488]: yes

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 3]
```diff
@@ -1393,104 +1393,118 @@ public function delete_question($sSessionKey, $iQuestionID)
...
+                "sid = :sid and parent_qid = 0 and title = :title",
+                array(
+                    ":sid" => $iSurveyID,
+                    ":title" => $sNewQuestionTitle
+                )
+            ));
+            if ($countQuestionTitle > 0) {
+                return array('status' => 'Error: Question title already exist in this survey.');
+            }
+            $importOptions = ['autorename' => true];
```
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/helpers/remotecontrol/remotecontrol_handle.php:1432]: It's to set autorename for https://github.com/LimeSurvey/LimeSurvey/blob/a0799596541616b97cdae779d852d335b93ae93f/application/helpers/admin/import_helper.php#L541

Maybe you're right `!empty($sNewQuestionTitle`) is a good idea. tomorrow

File: application/helpers/remotecontrol/remotecontrol_handle.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/helpers/remotecontrol/remotecontrol_handle.php:1485]: You mean create a  "Importing Question **text** through Remote Control doesn't set **help** for all languages"

[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/helpers/remotecontrol/remotecontrol_handle.php:1485]: That ticket is fixed by this PR.
Still, it is the same text that goes for all languages. We need to fix that later, on a FR

[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/helpers/remotecontrol/remotecontrol_handle.php:1485]: Ah OK :)
Yes : https://api.limesurvey.org/classes/remotecontrol_handle.html#method_import_question accept only string for `sNewqQuestion` and `sNewQuestionHelp`

**Comment by Shnoulle** [comment]
> I create the 5.X fix after , but need different fixes

Still waiting to fix it in master …

**Comment by Shnoulle** [comment]
> Code structure looks OK. Havne;t tested

You can find time to test ?

**Comment by gabrieljenik** [comment]
I would work on an automatic test for this.
Please, can you commit this to a LS branch so I can push also ?

**Comment by Shnoulle** [comment]
> I would work on an automatic test for this.

Oh ! OK , great :)

>  Please, can you commit this to a LS branch so I can push also ?

Yes , tomorrow :)

**Comment by gabrieljenik** [comment]
Please find here the test:
https://github.com/LimeSurvey/LimeSurvey/tree/bug/18071-importing-question-through-remote-control-may-fail-test-II

**Comment by Shnoulle** [comment]
Oups … sorry

**Comment by Shnoulle** [comment]
Arg … you set it for master … can not merge

**Pull request closed without merging**